### PR TITLE
Added Card data structures

### PR DESCRIPTION
**Pull request by simonmika**

### CODE REVIEW

File: Card/Expires/Month.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export type Month = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
+
+export namespace Month {
+	export function is(value: Month | any): value is Month {
+		return typeof value == "number" && Number.isInteger(value) && value <= 1 && value <= 12
```
Comment (qijieye) [inline Card/Expires/Month.ts:5]: `value <=  1` ?
it should be `value >= 1`?

File: Card/Expires/Year.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	| 94
+	| 95
+	| 96
+	| 97
+	| 98
+	| 99
+
+export namespace Year {
+	export function is(value: Year | any): value is Year {
+		return typeof value == "number" && Number.isInteger(value) && value <= 0 && value <= 99
```
Comment (qijieye) [inline Card/Expires/Year.ts:105]: `value >= 0`?

File: Card/Masked.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+	masked: string
+	expires: Expires
+}
+export namespace Masked {
+	export function is(value: Masked | any): value is Masked {
+		return (
+			typeof value == "object" &&
+			typeof value.pan == "string" &&
+			Expires.is(value.expires) &&
+			typeof value.csc == "string"
```
Comment (qijieye) [inline Card/Masked.ts:13]: `value.pan` & `value.csc`
are they correct for Masked interface?

File: Card/Expires/Month.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export type Month = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
```
Comment (lucy-meachem) [inline Card/Expires/Month.ts:1]: use a date library please

File: Card/Expires/index.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { Month as ExpiresMonth, Month } from "./Month"
```
Comment (lucy-meachem) [inline Card/Expires/index.ts:1]: date library please

File: Card/Masked.ts (hunk 3 above)
Comment (lucy-meachem) [inline Card/Masked.ts:13]: this function is (almost) identical to the one in Card, presumably there's an actual difference between the two? if you want to know if something is masked, surely you'd need to check the data not just its type

File: Card/index.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	export type Masked = CardMasked
+	export namespace Masked {
+		export const is = CardMasked.is
+	}
+	export type Token = string
+	export namespace Token {
+		export function is(value: any | Token): value is Token {
+			return (
+				typeof value == "string" &&
+				/[0-9]{10}:[0-2][0-9]{3}:[0-2][0-9][0-6][0,2,4,6,8]:[a-z,A-Z,0-9\-_]+:[a-z,A-Z,0-9\-_]+/.test(value)
```
Comment (lucy-meachem) [inline Card/index.ts:39]: document regexes like this please

File: Card/index.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+				expires: [
+					Number.parseInt(splitted[1].slice(0, 2)) as CardExpires.Month,
+					Number.parseInt(splitted[1].slice(2, 4)) as CardExpires.Year,
+				],
+			}
+		}
+		export function pack(card: Card, key: string, encrypted: string, salt: string): Token {
+			const length = card.pan.length
+			return [
+				card.pan.slice(0, 6) + card.pan.slice(length - 4, length),
```
Comment (lucy-meachem) [inline Card/index.ts:56]: why not just leave the *s in? saves future effort on the receiving end when visa introduce 8 digit bins

File: Card/index.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	}
+	export type Token = string
+	export namespace Token {
+		export function is(value: any | Token): value is Token {
+			return (
+				typeof value == "string" &&
+				/[0-9]{10}:[0-2][0-9]{3}:[0-2][0-9][0-6][0,2,4,6,8]:[a-z,A-Z,0-9\-_]+:[a-z,A-Z,0-9\-_]+/.test(value)
+			)
+		}
+		export function unpack(token: Token): CardMasked {
```
Comment (lucy-meachem) [inline Card/index.ts:42]: this function needs error handling; check lengths and formats

**Review by lucy-meachem — CHANGES_REQUESTED** [review]

**Review by ChriJ — COMMENTED** [review]
Nothing to add to what already was commented

**Pull request closed without merging**